### PR TITLE
Start of repo and AttachResult component

### PR DESCRIPTION
**Pull request by mikegron**
- Updated readme and other files, starting from search-ui-seed.
- Added AttachResult component and tests.

We want this repository to be home to generic and reusable components which could be used by any integrations in the future. So any comment is appreciated!

I started from [search-ui-seed](https://github.com/coveo/search-ui-seed) and added a basic, generic AttachResult component. The component is super simple and leverages Promises, it's basically a toggle button which also sends analytics events. You just have to give it 3 functions: attach, detach and optionally isAttached. These are configured just like any other component options, there is an example in index.html.

image image image

### CODE REVIEW

File: src/ui/AttachResult/AttachResult.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        detach: ComponentOptions.buildCustomOption(name => () => new Promise<void>((resolve, reject) => { resolve(); }), {
+            defaultFunction: () => () => new Promise<void>((resolve, reject) => { resolve(); })
+        })
+    };
+
+    constructor(
+        public element: HTMLElement,
+        public options: IAttachResultOptions,
+        public bindings?: IComponentBindings,
+        public queryResult?: IQueryResult
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: Does queryResult need to be a separate constructor arg, or does it make sense to put it under `options`?

File: src/ui/AttachResult/AttachResult.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        this.initialize();
+
+        this.bind.on(this.element, "click", this.toggleAttached);
+    }
+
+    /**
+     * Gets whether or not the result is currently attached.
+     */
+    public isAttached() : boolean {
+        return this.attached;
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: Could this return `undefined`? Would it need to be casted, for example by returning `return !!this.attached`?

File: src/ui/AttachResult/AttachResult.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * Attach the query result.
+     */
+    public attach() : Promise<void> {
+        if (this.attached || this.loading) {
+            return Promise.resolve();
+        }
+
+        this.setLoading(true);
+        return this.options.attach(this.queryResult)
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: I would recommend using `async/await` syntax rather than chaining `.then`

File: src/ui/AttachResult/AttachResult.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            .then(() => { this.attached = true; })
+            .then(() => { 
+                let customData: IAnalyticsCaseAttachMeta = {
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: It would be nice to extract this block as a separate method `logAttachCaseAnalyticsEvent`.

Similar comments for `detach`

File: src/ui/AttachResult/AttachResult.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                };
+          
+                this.usageAnalytics.logCustomEvent<IAnalyticsCaseDetachMeta>(analyticsActionCauseList.caseDetach, customData, this.root);
+            })
+            .finally(() => { this.setLoading(false); });
+    }
+
+    /** Toggle the state of the component. If the current result is not attached, attach it, if not, detach it. */
+    public toggleAttached() : void {
+        if (this.attached) {
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: For simple `if/else` logic like this one, I like using ternaries: e.g. 
```
this.attached ? this.detach() : this.attach()
```

File: src/ui/AttachResult/AttachResult.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+            this.detach();
+        } else {
+            this.attach();
+        }
+    }
+
+    protected initialize() : void {
+        // Adds HTML elements needed to display this component.
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: I would remove this comment. It is better to reserve them for "why" you are doing something rather than "what" you are doing.

File: src/ui/AttachResult/AttachResult.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        this.buttonElement = $$("div", {}, PaperclipIcon).el;
+        this.element.appendChild(this.buttonElement);
+
+        this.tooltipElement = $$("div", { className: "coveo-caption-for-icon" }).el;
+        this.element.appendChild(this.tooltipElement);
+
+        this.attached = false;
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: Could this be set on line 35 instead? If so, then my comment on casting using `!!` should no longer be necessary.

File: src/ui/AttachResult/AttachResult.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        this.render();
+
+        // Resolve the current result for the component and the initial state.
+        if (this.options.isAttached) {
```
Comment (samisayegh) [inline src/ui/AttachResult/AttachResult.ts]: I would extract this logic into a separate method, named something along the lines of `synchronizeAttachedState` or `updateAttachedState`

File: tests/ui/AttachResult.spec.ts [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+        it("should call isAttached on initialize", () => {
+            attachResult = Mock.optionsResultComponentSetup(AttachResult, { isAttached: faker.isAttachedMock }, fakeResult);
+            expect(faker.isAttachedMock).toHaveBeenCalledWith(fakeResult);
+        });
+
+        it("should be attached when isAttached returns true", (done) => {
+            isAttachedSpy.and.returnValue(Promise.resolve(true));
+            attachResult = Mock.optionsResultComponentSetup(AttachResult, { isAttached: faker.isAttachedMock }, fakeResult);
+            setTimeout(() => {
```
Comment (samisayegh) [inline tests/ui/AttachResult.spec.ts]: Seems strange that these `setTimeout` are needed?

File: tests/ui/AttachResult.spec.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+                detachCaption: "detach me",
+                attachCaption: "attach me"
+            }, fakeResult);
+
+            attachResult.cmp.attach().then(() => { done(); });
+        });
+
+        it("should only detach if attached", (done) => {
+            attachResult.cmp.detach()
+                .then(() => {
```
Comment (samisayegh) [inline tests/ui/AttachResult.spec.ts]: You could use async/await here,

File: src/ui/AttachResult/AttachResult.ts [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    attachCaption?: string;
+    /** Specifies the tooltip displayed when the result is already attached. */
+    detachCaption?: string;
+    /** The function that is called when the user wants to attach a result. */
+    attach: (queryResult: IQueryResult) => Promise<void>;
+    /** The function called when the user wants to un-link a result. */
+    detach: (queryResult: IQueryResult) => Promise<void>;
+    /** Optional function to check the initial state of the component. */
+    isAttached?: (queryResult: IQueryResult) => Promise<boolean>;
+}
```
Comment (ThibodeauJF) [inline src/ui/AttachResult/AttachResult.ts]: If you ever want to provide an option for tooltip positioning, instead of it always being at the bottom, for the Quickview component we use PopperJs
https://github.com/coveo/search-ui/blob/master/src/ui/Quickview/Quickview.ts

File: tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es5",
+    "noImplicitAny": false,
```
Comment (samisayegh) [inline tsconfig.json]: Since this is a young project, this would be a good option to set to `true`.

Other options that are nice: `noUnusedLocals`, `noImplicitReturns`

File: tslint.json [hunk 13]
```diff
@@ -0,0 +1,113 @@
+{
```
Comment (samisayegh) [inline tslint.json]: I recommend removing this config and using the opinionated linter [prettier ](https://www.npmjs.com/package/prettier) instead. As you prefer,

File: src/ui/AttachResult/AttachResult.ts (hunk 1 above)
Comment (erocheleau) [reply src/ui/AttachResult/AttachResult.ts]: What do you usually do with your components that require a result?

File: src/ui/AttachResult/AttachResult.ts [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        this.loading = loading;
+        this.render();
+    }
+
+    protected render() : void {
+        $$(this.buttonElement).removeClass("coveo-icon-attached");
+        $$(this.buttonElement).removeClass("coveo-icon-attach");
+        $$(this.buttonElement).removeClass("coveo-icon-loading");
+
+        if (this.loading) {
```
Comment (erocheleau) [inline src/ui/AttachResult/AttachResult.ts]: You could extract some of that logic into a private method: getClassFromState or something of the sort that would return either "coveo-icon-loading" or "coveo-icon-attached" or "coveo-icon-attach".

File: tests/ui/AttachResult.spec.ts (hunk 9 above)
Comment (erocheleau) [reply tests/ui/AttachResult.spec.ts]: To simulate the back and forth of a server call?

File: LICENSE [hunk 15]
```diff
@@ -1,201 +1,13 @@
-                                 Apache License
-                           Version 2.0, January 2004
-                        http://www.apache.org/licenses/
+Copyright 2017 Coveo
```
[the remark this answers is not in the document]
Comment (mikegron) [reply LICENSE]: yes it was asked by the sec team!

File: pages/index.html [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    <link rel="stylesheet" href="./css/CoveoFullSearch.css"/>
+    <link rel="stylesheet" href="./css/CoveoJsSearchExtensions.css"/>
+    <script src="http://localhost:8080/js/CoveoJsSearch.Lazy.js"></script>
+    <script src="js/CoveoJsSearchExtensions.js"></script>
+    <script>
+      document.addEventListener('DOMContentLoaded', function () {
+        Coveo.SearchEndpoint.configureSampleEndpointV2();
+        Coveo.init(document.body, {
+          AttachResult: {
+            attach: function(id) { return new Promise((resolve, reject) => { console.log("attached " + id); setTimeout(resolve, 1000);} ); },
```
[the remark this answers is not in the document]
Comment (mikegron) [reply pages/index.html]: good point!

File: sass/AttachResult.scss [hunk 17]
```diff
@@ -0,0 +1,71 @@
+$attachResult_grey: #adadad;
```
[the remark this answers is not in the document]
Comment (mikegron) [reply sass/AttachResult.scss]: I like this idea 👍

File: src/ui/AttachResult/AttachResult.ts (hunk 1 above)
Comment (mikegron) [reply src/ui/AttachResult/AttachResult.ts]: What i've seen is normally it's in the constructor but i'm not 100% why

File: tests/ui/AttachResult.spec.ts (hunk 9 above)
Comment (mikegron) [reply tests/ui/AttachResult.spec.ts]: i was having a hard time trying to make the test work, for some reason I cannot call `done()` in the spy `.and.callFake`, the test kept timeouting.. do you have an idea why?

File: src/ui/AttachResult/AttachResult.ts (hunk 1 above)
Comment (samisayegh) [reply src/ui/AttachResult/AttachResult.ts]: I see the `ChatterLikedBy` component in the search-ui accepts the `IQueryResult` as a fourth constructor param. What you did is consistent 🤷‍♂️

File: tests/ui/AttachResult.spec.ts (hunk 9 above)
Comment (samisayegh) [reply tests/ui/AttachResult.spec.ts]: Hmm, I've run into problems when using promises with the jest framework. There I solve it by moving the async method into the `beforeEach`. If the promise is the issue, maybe there is a thread that discusses it for Jasmine?

File: tests/ui/AttachResult.spec.ts (hunk 10 above)
Comment (mikegron) [reply tests/ui/AttachResult.spec.ts]: that's a lot cleaner :+1:

File: tsconfig.json (hunk 12 above)
Comment (mikegron) [reply tsconfig.json]: nice I didn't know these options

File: src/ui/AttachResult/AttachResult.ts (hunk 14 above)
Comment (mikegron) [reply src/ui/AttachResult/AttachResult.ts]: it's already kind of small and i still have to add the text so not sure it's worth it
